File: src/ladym/operations/recall.py
"""Two-tier retrieval with reflection gate (ARCHITECTURE.md §3).

Implements HyMem's cognitive-economy principle: a cheap tier-1 vector+activation pass answers
~70% of queries; only the ones that fail a cheap reflection check escalate to the expensive
tier-2 graph expansion + source backtracking.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

from ..config import Config, RecallConfig
from ..schema import Layer, Memory, MemoryType, RecallResponse, RecallResult
from ..storage.embeddings import EmbeddingProvider, tokenize
from ..storage.store import SQLiteStore
from .activation import (
    activation_score,
    infer_query_types,
    neighbour_counts_for,
)


@dataclass
class _ReflectionVerdict:
    sufficient: bool
    coverage: float
    n_hits: int


def _reflect(query: str, hits: list[RecallResult], cfg: RecallConfig) -> _ReflectionVerdict:
    """Cheap self-check: did tier-1 surface enough high-signal context?

    Default heuristic: count distinct query tokens covered by the union of hit contents, and
    the number of hits above a small similarity floor. Pluggable — ``Engine`` can substitute
    an LLM judge when one is configured.
    """
    q_tokens = set(tokenize(query)) - {"the", "a", "an", "is", "are", "to", "of", "and", "or"}
    if not q_tokens:
        return _ReflectionVerdict(sufficient=True, coverage=1.0, n_hits=len(hits))
    corpus = " ".join(r.memory.content + " " + r.memory.summary for r in hits).lower()
    covered = sum(1 for t in q_tokens if t in corpus)
    coverage = covered / len(q_tokens)
    sufficient = len(hits) >= cfg.reflection_min_hits and coverage >= cfg.reflection_min_coverage
    return _ReflectionVerdict(sufficient=sufficient, coverage=coverage, n_hits=len(hits))


def _rank(
    *,
    query_vec: list[float],
    candidates: list[tuple[Memory, float]],
    cfg: Config,
    neighbour_counts: dict[str, int],
    query_types: list[MemoryType],
) -> list[RecallResult]:
    """Apply the ACT-R activation function to candidates and return ranked results."""
    scored: list[tuple[float, Memory, float]] = []
    for mem, sim in candidates:
        act = activation_score(
            mem,
            query_similarity=sim,
            weights=cfg.activation,
            neighbour_counts=neighbour_counts_for([mem], neighbour_counts),
            query_types=query_types,
        )
        scored.append((act, mem, sim))
    scored.sort(key=lambda t: t[0], reverse=True)
    return [RecallResult(memory=m, score=act, tier=1, via=[]) for act, m, sim in scored]


def recall(
    store: SQLiteStore,
    embedder: EmbeddingProvider,
    query: str,
    *,
    cfg: Config,
    workspace: str | None = None,
    top_k: int | None = None,
    layers: list[Layer] | None = None,
    types: list[MemoryType] | None = None,
    min_similarity: float = 0.0,
) -> RecallResponse:
    """The full two-tier retrieval pipeline."""
    start = time.time()
    ws = workspace or cfg.workspace
    rcfg = cfg.recall
    k1 = top_k or rcfg.top_k_tier1
    query_vec = embedder.embed(query)
    query_types = types or infer_query_types(query)

    # ---- Tier 1: vector search + activation rerank ----
    raw_hits = store.vector_index.search(query_vec, top_k=max(k1 * 3, k1))
    cand: list[tuple[Memory, float]] = []
    for mem_id, sim in raw_hits:
        if sim < min_similarity:
            continue
        mem = store.get_memory(mem_id)
        if mem is None or mem.workspace != ws:
            continue
        if layers is not None and Layer(mem.layer) not in layers:
            continue
        if types is not None and MemoryType(mem.type) not in types:
            continue
        cand.append((mem, sim))

    neighbour_counts = store.associative_neighbour_counts() if hasattr(store, "associative_neighbour_counts") else {}

    tier1 = _rank(
        query_vec=query_vec,
        candidates=cand,
        cfg=cfg,
        neighbour_counts=neighbour_counts,
        query_types=query_types,
    )[:k1]

    verdict = _reflect(query, tier1, rcfg)
    if verdict.sufficient or not rcfg.enable_tier2:
        _commit_access(store, [r.memory.id for r in tier1])
        return RecallResponse(
            query=query,
            results=tier1,
            tier_reached=1,
            reflected_sufficient=verdict.sufficient,
            elapsed_ms=(time.time() - start) * 1000.0,
        )

    # ---- Tier 2: graph expansion + source backtracking ----
    expanded = _tier2_expand(store, tier1, cfg, ws)
    # merge: tier-2 results inherit tier=2; original tier1 items keep tier=1
    by_id: dict[str, RecallResult] = {r.memory.id: r for r in tier1}
    for mem, sim, via in expanded:
        if mem.id in by_id:
            continue
        if layers is not None and Layer(mem.layer) not in layers:
            continue
        if types is not None and MemoryType(mem.type) not in types:
            continue
        act = activation_score(
            mem,
            query_similarity=sim,
            weights=cfg.activation,
            neighbour_counts=neighbour_counts_for([mem], neighbour_counts),
            query_types=query_types,
        )
        by_id[mem.id] = RecallResult(memory=mem, score=act, tier=2, via=via)
    merged = sorted(by_id.values(), key=lambda r: r.score, reverse=True)[: (top_k or rcfg.top_k_tier2)]

    _commit_access(store, [r.memory.id for r in merged])
    return RecallResponse(
        query=query,
        results=merged,
        tier_reached=2,
        reflected_sufficient=True,
        elapsed_ms=(time.time() - start) * 1000.0,
    )


def _tier2_expand(
    store: SQLiteStore,
    tier1: list[RecallResult],
    cfg: Config,
    workspace: str,
) -> list[tuple[Memory, float, list[str]]]:
    """Walk the associative graph from tier-1 anchors, plus backtrack to source context."""
    out: list[tuple[Memory, float, list[str]]] = []
    seen: set[str] = {r.memory.id for r in tier1}
    qvec_cache: dict[str, list[float]] = {}

    for anchor in tier1:
        frontier = [(anchor.memory.id, 1, [anchor.memory.id])]
        while frontier:
            cur_id, depth, path = frontier.pop(0)
            if depth > cfg.recall.graph_hops:
                continue
            for edge in store.neighbors(cur_id):
                other_id = edge.dst_id if edge.src_id == cur_id else edge.src_id
                if other_id in seen:
                    continue
                seen.add(other_id)
                other = store.get_memory(other_id)
                if other is None or other.workspace != workspace:
                    continue
                # similarity to query, lazily embedding
                if other_id not in qvec_cache:
                    qvec_cache[other_id] = [0.0]  # placeholder; we don't have query vec here
                out.append((other, max(0.05, anchor.score * 0.5 / depth), path + [other_id]))
                frontier.append((other_id, depth + 1, path + [other_id]))

    # backtrack: for code symbols in tier1/expanded, pull their file memory too
    for mem_id in list(seen):
        mem = store.get_memory(mem_id)
        if mem and mem.type == MemoryType.CODE_SYMBOL.value:
            file_meta = mem.metadata.get("file_path")
            if not file_meta:
                continue
            for m in store.iter_memories(workspace=workspace, type_=MemoryType.CODE_FILE.value):
                if m.metadata.get("file_path") == file_meta and m.id not in seen:
                    seen.add(m.id)
                    out.append((m, 0.1, [mem_id, m.id]))
    return out


def _commit_access(store: SQLiteStore, ids: list[str]) -> None:
    """Side-effect: bump access_count / last_access_at on recalled items (drives decay)."""
    now = time.time()
    for mid in ids:
        store.touch_memory(mid, now=now)

File: operations/recall.go
package operations

import (
	"sort"
	"strings"
	"time"

	"github.com/ProjAnvil/LadyM/config"
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

type reflectionVerdict struct {
	sufficient bool
	coverage   float64
	nHits      int
}

var stopwords = map[string]bool{
	"the": true, "a": true, "an": true, "is": true, "are": true,
	"to": true, "of": true, "and": true, "or": true,
}

func reflect(query string, hits []*schema.RecallResult, cfg config.RecallConfig) reflectionVerdict {
	qTokens := map[string]bool{}
	for _, t := range storage.Tokenize(query) {
		if !stopwords[t] {
			qTokens[t] = true
		}
	}
	if len(qTokens) == 0 {
		return reflectionVerdict{sufficient: true, coverage: 1.0, nHits: len(hits)}
	}
	var corpus strings.Builder
	for _, r := range hits {
		corpus.WriteString(r.Memory.Content)
		corpus.WriteString(" ")
		corpus.WriteString(r.Memory.Summary)
	}
	lower := strings.ToLower(corpus.String())
	covered := 0
	for t := range qTokens {
		if strings.Contains(lower, t) {
			covered++
		}
	}
	coverage := float64(covered) / float64(len(qTokens))
	sufficient := len(hits) >= cfg.ReflectionMinHits && coverage >= cfg.ReflectionMinCoverage
	return reflectionVerdict{sufficient: sufficient, coverage: coverage, nHits: len(hits)}
}

func rank(candidates []candidate, cfg *config.Config, neighbourCounts map[string]int, queryTypes []schema.MemoryType) []*schema.RecallResult {
	out := make([]*schema.RecallResult, 0, len(candidates))
	for _, c := range candidates {
		act := ActivationScore(c.Memory, c.Sim, cfg.Activation, neighbourCounts, queryTypes, 0)
		out = append(out, &schema.RecallResult{Memory: c.Memory, Score: act, Tier: 1, Via: []string{}})
	}
	sort.SliceStable(out, func(a, b int) bool { return out[a].Score > out[b].Score })
	return out
}

type candidate struct {
	Memory *schema.Memory
	Sim    float64
}

// Recall runs the full two-tier retrieval pipeline.
func Recall(store *storage.SQLiteStore, embedder storage.EmbeddingProvider, query string, cfg *config.Config, workspace string, topK int, layers []schema.Layer, types []schema.MemoryType, minSimilarity float64) (*schema.RecallResponse, error) {
	start := time.Now()
	ws := workspace
	if ws == "" {
		ws = cfg.Workspace
	}
	rcfg := cfg.Recall
	k1 := topK
	if k1 == 0 {
		k1 = rcfg.TopKTier1
	}
	queryVec, err := embedder.Embed(query)
	if err != nil {
		return nil, err
	}
	queryTypes := types
	if len(queryTypes) == 0 {
		queryTypes = InferQueryTypes(query)
	}

	fetchK := k1 * 3
	if fetchK < k1 {
		fetchK = k1
	}
	rawHits := store.VectorIndex().Search(queryVec, fetchK)
	var cand []candidate
	for _, h := range rawHits {
		if h.Similarity < minSimilarity {
			continue
		}
		mem, err := store.GetMemory(h.ID)
		if err != nil {
			return nil, err
		}
		if mem == nil || mem.Workspace != ws {
			continue
		}
		if IsRetired(mem) {
			continue
		}
		if layers != nil && !layerIn(mem.Layer, layers) {
			continue
		}
		if types != nil && !typeIn(mem.Type, types) {
			continue
		}
		cand = append(cand, candidate{Memory: mem, Sim: h.Similarity})
	}

	neighbourCounts, err := store.NeighborCounts()
	if err != nil {
		return nil, err
	}

	tier1 := rank(cand, cfg, neighbourCounts, queryTypes)
	if len(tier1) > k1 {
		tier1 = tier1[:k1]
	}

	verdict := reflect(query, tier1, rcfg)
	if verdict.sufficient || !rcfg.EnableTier2 {
		commitAccess(store, resultIDs(tier1))
		return &schema.RecallResponse{
			Query: query, Results: tier1, TierReached: 1,
			ReflectedSufficient: verdict.sufficient,
			ElapsedMs:           float64(time.Since(start).Milliseconds()),
		}, nil
	}

	expanded, err := tier2Expand(store, tier1, cfg, ws)
	if err != nil {
		return nil, err
	}
	byID := map[string]*schema.RecallResult{}
	for _, r := range tier1 {
		byID[r.Memory.ID] = r
	}
	for _, ex := range expanded {
		if _, ok := byID[ex.Memory.ID]; ok {
			continue
		}
		if layers != nil && !layerIn(ex.Memory.Layer, layers) {
			continue
		}
		if types != nil && !typeIn(ex.Memory.Type, types) {
			continue
		}
		act := ActivationScore(ex.Memory, ex.Sim, cfg.Activation, neighbourCounts, queryTypes, 0)
		byID[ex.Memory.ID] = &schema.RecallResult{Memory: ex.Memory, Score: act, Tier: 2, Via: ex.Via}
	}
	k2 := topK
	if k2 == 0 {
		k2 = rcfg.TopKTier2
	}
	merged := make([]*schema.RecallResult, 0, len(byID))
	for _, r := range byID {
		merged = append(merged, r)
	}
	sort.SliceStable(merged, func(a, b int) bool { return merged[a].Score > merged[b].Score })
	if len(merged) > k2 {
		merged = merged[:k2]
	}

	commitAccess(store, resultIDs(merged))
	return &schema.RecallResponse{
		Query: query, Results: merged, TierReached: 2,
		ReflectedSufficient: true,
		ElapsedMs:           float64(time.Since(start).Milliseconds()),
	}, nil
}

func layerIn(l schema.Layer, layers []schema.Layer) bool {
	for _, x := range layers {
		if l == x {
			return true
		}
	}
	return false
}

func typeIn(t schema.MemoryType, types []schema.MemoryType) bool {
	for _, x := range types {
		if t == x {
			return true
		}
	}
	return false
}

func resultIDs(results []*schema.RecallResult) []string {
	ids := make([]string, 0, len(results))
	for _, r := range results {
		ids = append(ids, r.Memory.ID)
	}
	return ids
}

type expandedItem struct {
	Memory *schema.Memory
	Sim    float64
	Via    []string
}

func tier2Expand(store *storage.SQLiteStore, tier1 []*schema.RecallResult, cfg *config.Config, workspace string) ([]expandedItem, error) {
	var out []expandedItem
	seen := map[string]bool{}
	for _, r := range tier1 {
		seen[r.Memory.ID] = true
	}

	type frontierNode struct {
		id    string
		depth int
		path  []string
	}

	for _, anchor := range tier1 {
		frontier := []frontierNode{{anchor.Memory.ID, 1, []string{anchor.Memory.ID}}}
		for len(frontier) > 0 {
			cur := frontier[0]
			frontier = frontier[1:]
			if cur.depth > cfg.Recall.GraphHops {
				continue
			}
			// follow supersedes to the newest version even if filtered from tier-1
			supEdges, err := store.Neighbors(cur.id, "supersedes")
			if err != nil {
				return nil, err
			}
			for _, e := range supEdges {
				if e.SrcID == cur.id && !seen[e.DstID] {
					seen[e.DstID] = true
					newer, err := store.GetMemory(e.DstID)
					if err != nil {
						return nil, err
					}
					if newer != nil && newer.Workspace == workspace {
						out = append(out, expandedItem{Memory: newer, Sim: maxF(0.05, anchor.Score*0.6/float64(cur.depth)), Via: append(append([]string{}, cur.path...), e.DstID)})
					}
				}
			}
			edges, err := store.Neighbors(cur.id, "")
			if err != nil {
				return nil, err
			}
			for _, e := range edges {
				otherID := e.DstID
				if e.SrcID == cur.id {
					otherID = e.DstID
				} else {
					otherID = e.SrcID
				}
				if seen[otherID] {
					continue
				}
				seen[otherID] = true
				other, err := store.GetMemory(otherID)
				if err != nil {
					return nil, err
				}
				if other == nil || other.Workspace != workspace {
					continue
				}
				path := append(append([]string{}, cur.path...), otherID)
				out = append(out, expandedItem{Memory: other, Sim: maxF(0.05, anchor.Score*0.5/float64(cur.depth)), Via: path})
				frontier = append(frontier, frontierNode{otherID, cur.depth + 1, path})
			}
		}
	}

	// backtrack: for code symbols, pull their file memory too
	for memID := range seen {
		mem, err := store.GetMemory(memID)
		if err != nil {
			return nil, err
		}
		if mem == nil || mem.Type != schema.TypeCodeSymbol {
			continue
		}
		filePath := mem.MetaString("file_path")
		if filePath == "" {
			continue
		}
		files, err := store.IterMemories(workspace, "", string(schema.TypeCodeFile))
		if err != nil {
			return nil, err
		}
		for _, m := range files {
			if m.MetaString("file_path") == filePath && !seen[m.ID] {
				seen[m.ID] = true
				out = append(out, expandedItem{Memory: m, Sim: 0.1, Via: []string{memID, m.ID}})
			}
		}
	}
	return out, nil
}

func maxF(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

func commitAccess(store *storage.SQLiteStore, ids []string) {
	now := schema.Now()
	for _, id := range ids {
		_ = store.TouchMemory(id, now)
	}
}
